lib/src/ui/pages/authentication/registration/registration_notifier.dart

```diff
@@ -0,0 +1,93 @@
+import 'package:allthenews/generated/l10n.dart';
+import 'package:allthenews/src/domain/authorization/authentication_repository.dart';
+import 'package:allthenews/src/domain/communication/exception_mapper.dart';
+import 'package:allthenews/src/ui/pages/authentication/registration/registration_state.dart';
+import 'package:flutter/cupertino.dart';
+
+class RegistrationNotifier extends ChangeNotifier {
+  final AuthenticationRepository _authorizationRepository;
+  final ExceptionMapper _exceptionMapper;
+
+  RegistrationNotifier(this._authorizationRepository, this._exceptionMapper);
+
+  RegistrationState _state = const RegistrationState();
+
+  RegistrationState get state => _state;
+
+  VoidCallback returnToProfile;
+
+  Future<void> createUser(String email, String password, String name) async {
+    _validateFields();
+
+    if (!_state.canSubmit) {
+      notifyListeners();
+      return;
+    }
+
+    _setNotifierState(_state.copyWith(isLoading: true));
+
+    try {
+      await _authorizationRepository.createUser(email, password);
+      _updateUserName(name);
+    } on Exception catch (exception) {
+      _setNotifierState(_state.copyWith(
+        exception: _exceptionMapper.toDomainException(exception),
```

Comment:
mapowanie powinno odbyć się w warstwie data, gdzieś bliżej komunikacji, w tym przypadku w firebaseAuthenticationRepository

Reply:
właśnie moim zdaniem powinny być mapowane jak najbliżej warstwy prezentacji, żeby nie przepychać tych błędów przez wszystkie warstwy, ale tak jak napisałem we wcześniejszym komentarzu

```
Co do exceptionów to myślę, że dobrze byłoby ten temat przegadać i uspójnić, celowo dałem to domeny, bo w ten sposób przy wielu warstwach tylko w jednym miejscu trzeba łapać wyjątek, a nie w każdej warstwie po kolei go przepychać,
```

Reply:
Generalnie exceptiony są domeną. W warstwie prezentacji powinien być jakiś mapper, który na podstawie błędów domenowych wyświetla odpowiedni komunikat. Tak jak zwrotka w postaci sukcesu jest mapowana na obiekt domenowy np. Article, tak też wszystkie błędy powinny być mapowane w warstwie danych tak, aby warstwa domeny nie miała żadnych zależności komunikacyjnych, ani żeby błędy komunikacyjne nie przebijały się do warstwy prezentacji. Data korzysta z domeny, ui korzysta z domeny i to tyle z zależności między warstwami. Wiesz, skoro jsona mapujemy na domenę to czemu nie error? W przypadku przepychania błędów z api wprost do ui, domena i ui przechowują te błędy. I przykładowo dostaniemy DioError, czemu w ogóle UI ma wiedzieć, że korzystamy z Dio jako klienta http, albo z firebase? Ucinanie tego poprzez mapowanie w data powoduje brak konieczności wprowadzania zmian w pozostałych warstwach.

Reply:
czyli wychodzi na to, że jedynym miejscem gdzie nie będzie mapowań jest właśnie domena. Tak to w data będzie mapowania błędów api/firebase czy tam cokolwiek innego na domenowe błędy, a w warstwie ui domenowe błędy na stringsy

Reply:
Tak :) Analogicznie mapujemy teraz jeżeli chodzi o Article, w data z jsona na domenę, a w presentation z domeny na jakieś viewEntity w viewState :)